### PR TITLE
[Master] Add getUser function to userSessionStore to retrieve authenticated user by user Id

### PR DESCRIPTION
**Pull request by Thumimku**
Public Issue: https://github.com/wso2/product-is/issues/24066
APIM Branched Public PR: https://github.com/wso2/carbon-identity-framework/pull/5455

This pull request adds the ability to retrieve detailed user information, including the associated Identity Provider (IDP), from the user session store by user ID. It introduces a new method in `UserSessionStore`, updates SQL queries, and extends the test coverage to ensure correct behavior and error handling for this feature.

**Enhancements to User Retrieval:**

* Added a new method `getUser(String userId)` in `UserSessionStore` to fetch user details and associated IDP using the user ID. This method validates input, queries the database, and maps results to an `AuthenticatedUser` object.
* Introduced a new SQL query `SQL_SELECT_USER_FROM_USER_ID` in `SQLQueries` to join `IDN_AUTH_USER` and `IDP` tables and retrieve user and IDP details by user ID.

**Testing Improvements:**

* Expanded `UserSessionStoreTest` to include tests for the new `getUser` method, covering successful retrieval, invalid user IDs, and null input scenarios. Also added a data provider for federated user test cases. [[1]](diffhunk://#diff-6be6e3d17c2f2f9fb1753c35f4bd8aa07243a42a307452696058dfe946920922R105-R113) [[2]](diffhunk://#diff-6be6e3d17c2f2f9fb1753c35f4bd8aa07243a42a307452696058dfe946920922R224-R255)
* Updated test imports to include `AuthenticatedUser` for the new test cases.

**Test Database Schema Updates:**

* Added the `IDP` table and a sample "FEDERATED" IDP entry to the H2 test database schema to support join queries in tests.

## New Features
* Added user lookup functionality that retrieves user details including name, tenant information, domain, and associated identity provider details by user ID, with built-in input validation.

## Tests
* Added comprehensive test coverage for the new user lookup feature, validating successful user retrieval, null input error handling, and federated authentication scenarios.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 1]
```diff
@@ -1334,4 +1335,38 @@ public String getFederatedUserId(String subjectIdentifier, int tenantId, int idP
...
     }
+
+    /**
+     * Method to retrieve user and associated IDP available in the IDN_AUTH_USER table.
+     *
+     * @param userId Id of the authenticated user
+     * @return the user and associated IDP
+     * @throws UserSessionException if an error occurs when retrieving the mapping from the database
+     */
+    public AuthenticatedUser getUser(String userId) throws UserSessionException {
```
[the remark this answers is not in the document]
Comment (Thumimku) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:1346]: False positive

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 2]
```diff
@@ -1334,4 +1335,38 @@ public String getFederatedUserId(String subjectIdentifier, int tenantId, int idP
...
+
+        if (StringUtils.isBlank(userId)) {
+            throw new UserSessionException("Invalid userId: userId cannot be null or empty.");
+        }
```
[the remark this answers is not in the document]
Comment (Thumimku) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:1350]: No need to log and throw

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 3]
```diff
@@ -1334,4 +1335,38 @@ public String getFederatedUserId(String subjectIdentifier, int tenantId, int idP
...
+                        user.setTenantDomain(IdentityTenantUtil.getTenantDomain(resultSet.getInt(2)));
+                        user.setUserStoreDomain(resultSet.getString(3));
+                        user.setFederatedIdPName(resultSet.getString(4));
+                    }
+                }
+            }
+            return user;
+        } catch (SQLException e) {
+            throw new UserSessionException("Error while retrieving information of user id: " + userId, e);
+        }
```
[the remark this answers is not in the document]
Comment (Thumimku) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:1370]: no need to log and throw

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/dbScripts/user_session_store_h2.sql [hunk 4]
```diff
@@ -45,3 +45,35 @@ CREATE TABLE IF NOT EXISTS IDN_AUTH_SESSION_APP_INFO (
...
+  IS_LOCAL_CLAIM_DIALECT     CHAR(1)      NOT NULL DEFAULT '0',
+  DISPLAY_NAME               VARCHAR(255),
+  IMAGE_URL                  VARCHAR(1024),
+  UUID                       CHAR(36) NOT NULL,
+  PRIMARY KEY (ID),
+  UNIQUE (TENANT_ID, NAME),
+  UNIQUE (UUID)
+);
+
+-- IDP entry for FEDERATED idp for user session store tests.
```
[the remark this answers is not in the document]
Comment (Thumimku) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/dbScripts/user_session_store_h2.sql:77]: False positive
